consensus/core/src/tests/randomized_tests.rs

```diff
@@ -0,0 +1,233 @@
+#[test]
+fn test_randomized_dag_all_direct_commit() {
+    let random_test_setup = random_test_setup();
+    let authority = authority_setup(random_test_setup.num_authorities);
+
+    let pipeline = random_test_setup.num_waves % DEFAULT_WAVE_LENGTH as usize;
+    let wave_number =
+        authority.committer.committers[pipeline].wave_number(random_test_setup.num_waves as u32);
+    let num_rounds = authority.committer.committers[pipeline].decision_round(wave_number);
+    let include_leader_percentage = 100;
+    let dag_builder = create_random_dag(
+        random_test_setup.seed,
+        include_leader_percentage,
+        num_rounds,
+        authority.context.clone(),
+    );
+
+    dag_builder.persist_all_blocks(authority.dag_state.clone());
+
+    tracing::info!(
+        "Running test with committee size {} & {} completed waves in the DAG...",
+        random_test_setup.num_authorities,
+        random_test_setup.num_waves
+    );
+
+    let last_decided = Slot::new_for_test(0, 0);
+    let sequence = authority.committer.try_decide(last_decided);
+    tracing::debug!("Commit sequence: {sequence:#?}");
+
+    assert_eq!(sequence.len(), random_test_setup.num_waves);
+    for (i, leader_block) in sequence.iter().enumerate() {
+        // First sequenced leader should be in round 1.
+        let leader_round = i as u32 + 1;
+        if let DecidedLeader::Commit(ref block) = leader_block {
+            assert_eq!(block.round(), leader_round);
+            assert_eq!(
+                block.author(),
+                authority.committer.get_leaders(leader_round)[0]
+            );
+        } else {
+            panic!("Expected a committed leader")
+        };
+    }
+}
+
+/// Test builds a randomized dag with the following conditions:
+/// - Links to 2f+1 minimum ancestors
+/// - Links to leader of previous round 50% of the time.
+///
+/// Blocks will randomly be fed through BlockManager and after each accepted
+/// block we will try_decide() and if there is a committed sequence we will update
+/// last_decided and continue. We do this from the perspective of two different
+/// authorities who receive the blocks in different orders and ensure the resulting
+/// sequence is the same for both authorities. The resulting sequence will include
+/// Commit & Skip decisions and potentially will stop before coming to a decision
+/// on all waves as we may have an Undecided leader somewhere early in the sequence.
+#[test]
+fn test_randomized_dag_and_decision_sequence() {
+    let mut random_test_setup = random_test_setup();
+
+    // Setup for Authority 1
+    let mut authority_1 = authority_setup(random_test_setup.num_authorities);
+
+    let pipeline = random_test_setup.num_waves % DEFAULT_WAVE_LENGTH as usize;
+    let wave_number =
+        authority_1.committer.committers[pipeline].wave_number(random_test_setup.num_waves as u32);
+    let num_rounds = authority_1.committer.committers[pipeline].decision_round(wave_number);
```

Comment:
Is it possible to hide the details of calculating num_rounds? Maybe we can just pick a constant max round across tests, because using less rounds seems not increasing the test coverage.